### DISCUSSION TITLE
HyperlinkButton setting visited link color doesn't work

### DISCUSSION THREAD

**Discussion by TangDave** [AvaloniaUI/Avalonia#17002] [Q&A] [answered]
(Avalonia 11.1.2 on Windows) I'm doing the following and expecting it to affect the visited color of a HyperlinkButton:
```
		<Style Selector="HyperlinkButton:visited /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Foreground" Value="{DynamicResource LinkVisitedColor}"/>
		</Style>
```
However, this doesn't have any effect on the visited link color.

**Comment by stevemonaco** [comment]
Make sure this is a brush. Assuming this is SimpleTheme, you may wish to replace the `HyperlinkVisitedBrush` keyed resource instead of writing a selector. It's called `HyperlinkButtonForegroundVisited` in Fluent theme.

Also, ensure `LinkVisitedColor` is a brush.

**Reply by TangDave** [reply]
Yes, LinkVisitedColor is a SolidColorBrush - I've defined it as:
`<SolidColorBrush x:Key='LinkVisitedColor'>Indigo</SolidColorBrush>`

For dark theme, I'm trying to see other colors which work better - the default doesn't cut it.

How would I replace the HyperlinkVisitedBrush keyed resource?

**Reply by stevemonaco** [reply]
Anywhere that supports `<X.Resources>` (eg. App or any control). I typically use create a separate `ResourceDictionary` especially for replacements and then `ResourceInclude` them in App.

```xml
<UserControl.Resources>
    <SolidColorBrush x:Key="HyperlinkVisitedBrush" Color="Indigo" />
</UserControl.Resources>
```

**Reply by TangDave** [reply]
I can't seem to add that anywhere without errors.

Here's a more complete picture of what I have in my App.axaml file, with only the HyperlinkButton colors included:

```
	<Application.Resources>
		<ResourceDictionary>
			<ResourceDictionary.ThemeDictionaries>
				
				<ResourceDictionary x:Key='Light'>
					<SolidColorBrush x:Key='LinkColor'>MediumBlue</SolidColorBrush>
					<SolidColorBrush x:Key='LinkHoverColor'>FireBrick</SolidColorBrush>
					<SolidColorBrush x:Key='LinkVisitedColor'>Indigo</SolidColorBrush>
				</ResourceDictionary>

				<ResourceDictionary x:Key='Dark'>
					<SolidColorBrush x:Key='LinkColor'>CadetBlue</SolidColorBrush>
					<SolidColorBrush x:Key='LinkHoverColor'>IndianRed</SolidColorBrush>
					<SolidColorBrush x:Key='LinkVisitedColor'>MediumPurple</SolidColorBrush>
				</ResourceDictionary>

			</ResourceDictionary.ThemeDictionaries>
		</ResourceDictionary>
	</Application.Resources>

	<Application.Styles>
		<FluentTheme/>

		<Style Selector="HyperlinkButton">
			<Setter Property="Foreground" Value="{DynamicResource LinkColor}"/>
		</Style>

		<Style Selector="HyperlinkButton:pointerover /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Foreground" Value="{DynamicResource LinkHoverColor}"/>
		</Style>

		<Style Selector="HyperlinkButton:visited /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Foreground" Value="{DynamicResource LinkVisitedColor}"/>
		</Style>

	</Application.Styles>
```
Everything except the 'visited' color works fine.

**Reply by stevemonaco** [reply, marked as answer]
I tried to repro with 11.2.0-preview1 and it seems to work fine. I changed the visit color to Orange. Finding a difference between the default purple visited shade and the shade of purple you picked was hard.

![avalonia-linkvisit](https://github.com/user-attachments/assets/5c75af51-2fee-48db-9323-1df60a9333ab)

Since you already have theme-specific use setup, it's easy to replace keyed resources. You're using Fluent, so the following works (remove all selectors):

```xml
<ResourceDictionary x:Key="Dark">
    <SolidColorBrush x:Key="HyperlinkButtonForeground">CadetBlue</SolidColorBrush>
    <SolidColorBrush x:Key="HyperlinkButtonForegroundPointerOver">IndianRed</SolidColorBrush>
    <SolidColorBrush x:Key="HyperlinkButtonForegroundVisited">Orange</SolidColorBrush>
</ResourceDictionary>
```

**Reply by TangDave** [reply]
I guess it's a bug in 11.1.2.
Your example works fine, *except* for the HyperlinkButtonForegroundVisited brush.
